### ISSUE TITLE
Multiple callbacks in a Future

### ISSUE THREAD

**Issue by funny-falcon** [nim-lang/Nim#5200]
It is quite strange that Future may have only one callback. It limits its composability.

**Comment by Araq** [comment]
Why? Shouldn't the callback "compose" the stuff?

**Comment by funny-falcon** [comment]
No, cause many libraries works only with default callback, and they will ovewrite "composing" callback.

**Comment by krux02** [comment]
Can you give an example on what you would like to be able to do?

**Comment by funny-falcon** [comment]
For example, i would like to store a future in a table by a key, and have many consumers to be informed when future will be completed.
And I want it to be transparently supported with all standard library.

In simple words, I want Javascript Promise or Dart's Future or C# Task:
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then
- https://api.dartlang.org/stable/1.21.0/dart-async/Future/then.html

They all may have multiple consumers of result. It is really useful feature, and it will be quite painful to implement outside of standard library

**Comment by krux02** [comment]
What you want is the ``Future`` to have an error callback. Then you should say so in the title.

**Comment by funny-falcon** [comment]
No, I didn't mean error callback. Error callback is not as important as multiple callbacks.
Error could be signaled with special result value.
Mutiple callbacks are much more useful/important.

**Comment by funny-falcon** [comment]
Promise#then and Future#then could be called several times, and all passed callbacks will be executed.

Any way, I'm not Nim's user (and Future's single callback is one of reasons), so you may close the issue.

But I bet, it will be raised again by other users.

**Comment by Araq** [comment]
> Any way, I'm not Nim's user (and Future's single callback is one of reasons), so you may close the issue.

Don't worry, that's not how things work here. :-)

**Comment by dom96** [comment]
I'll have to think about this more but my initial thoughts is that this is a good idea.

**Comment by andreaferretti** [comment]
Related: https://github.com/nim-lang/Nim/issues/5731
